WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs

```diff
@@ -32,18 +32,30 @@ public LabelSelectionViewModel(Money targetAmount)
 
 	public IEnumerable<LabelViewModel> LabelsBlackList => AllLabelsViewModel.Where(x => x.IsBlackListed);
 
-	public Pocket[] AutoSelectPockets()
+	public Pocket[] AutoSelectPockets(SmartLabel recipient)
 	{
 		var knownPockets = NonPrivatePockets.Where(x => x.Labels != CoinPocketHelper.UnlabelledFundsText).ToArray();
 		var unknownPockets = NonPrivatePockets.Except(knownPockets).ToArray();
 		var privateAndUnknownPockets = _allPockets.Except(knownPockets).ToArray();
 		var privateAndKnownPockets = _allPockets.Except(unknownPockets).ToArray();
+		var knownByRecipientPockets = knownPockets.Where(pocket => pocket.Labels.Any(recipient.Contains)).ToArray();
```

Comment:
AFAIU this is not correct.
If the recipient is `"David", "Lucas"` then we should not select coins that have only `"David"` or `"Lucas"`.

If the above is correct this is how I would correct the test:

```
public void AutoSelectKnownByRecipientPockets()
{
	var selection = new LabelSelectionViewModel(Money.Parse("1.0"));

	var pockets = new List<Pocket>();
	pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
	pockets.AddPocket(1.1M, out var pocket2, "Dan");
	pockets.AddPocket(0.5M, out var pocket3, CoinPocketHelper.UnlabelledFundsText);
	pockets.AddPocket(1.1M, out var pocket4, "David", "Lucas");
	pockets.AddPocket(1.1M, out var pocket5, "David");
	pockets.AddPocket(1.1M, out var pocket6, "Lucas");
	pockets.AddPocket(1.1M, out var pocket7, "David", "Lucas", "Dan");

	selection.Reset(pockets.ToArray());

	var output = selection.AutoSelectPockets(new SmartLabel("David", "Lucas"));
	Assert.DoesNotContain(pocket1, output);
	Assert.DoesNotContain(pocket2, output);
	Assert.DoesNotContain(pocket3, output);
	Assert.Contains(pocket4, output);
	Assert.DoesNotContain(pocket5, output);
	Assert.DoesNotContain(pocket6, output);
	Assert.Contains(pocket7, output);
}
```

Reply:
Why? If we don't have coins with `David, Lucas` then it is better to use `David` or `Lucas` instead of a 3rd like `Dan`.

Reply:
The goal with this line:
```
var knownByRecipientPockets = knownPockets.Where(pocket => pocket.Labels.Any(recipient.Contains)).ToArray();
```
was to get all the pockets that contain ANY of the labels from the recipient list. What is the point of this? Better to use these pockets instead of those that have no relation at all with the recipients. Rename the variable to `knownByAnyRecipientPockets` maybe?

If there is a pocket with the exact match `onlyKnownByRecipientPockets` handles it.

What do you think about sorting the `knownByRecipientPockets` by the number of the labels that are found in the recipient?

So if the recipient is `David, Lucas`... and there are the pockets of `David`, `Lucas`, `David, Lucas, Dan` after sorting the `David, Lucas, Dan` will be in the first place so the transaction would use that.

Regarding you example:
```
public void AutoSelectKnownByRecipientPockets()
{
	var selection = new LabelSelectionViewModel(Money.Parse("1.0"));

	var pockets = new List<Pocket>();
	pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
	pockets.AddPocket(1.1M, out var pocket2, "Dan");
	pockets.AddPocket(0.5M, out var pocket3, CoinPocketHelper.UnlabelledFundsText);
	pockets.AddPocket(1.1M, out var pocket4, "David", "Lucas");
	pockets.AddPocket(1.1M, out var pocket5, "David");
	pockets.AddPocket(1.1M, out var pocket6, "Lucas");
	pockets.AddPocket(1.1M, out var pocket7, "David", "Lucas", "Dan");

	selection.Reset(pockets.ToArray());

	var output = selection.AutoSelectPockets(new SmartLabel("David", "Lucas"));
	Assert.DoesNotContain(pocket1, output);
	Assert.DoesNotContain(pocket2, output);
	Assert.DoesNotContain(pocket3, output);
	Assert.Contains(pocket4, output);
	Assert.DoesNotContain(pocket5, output);
	Assert.DoesNotContain(pocket6, output);
	Assert.Contains(pocket7, output);
}
```

This should fail. Because the goal is `David, Lucas` and there is a pocket with the exact same labels which covers the `1 BTC` amount. Pocket 7 should not be included.

Reply:
> This should fail. Because the goal is David, Lucas and there is a pocket with the exact same labels which covers the 1 BTC amount. Pocket 7 should not be included.

That is true. This change should make it pass `pockets.AddPocket(0.9M, out var pocket4, "David", "Lucas");`

> The goal with this line:
> 
> var knownByRecipientPockets = knownPockets.Where(pocket => pocket.Labels.Any(recipient.Contains)).ToArray();
> was to get all the pockets that contain ANY of the labels from the recipient list. What is the point of this? Better to use these pockets instead of those that have no relation at all with the recipients. 

That line still can select pockets that have labels without any relation to the recipient, for example `David, Dan`.

Let's say the recipient is `David, Lucas` and you want to send 1 btc and you have the following coins:

coin1: 0.5 btc | "David"
coin2: 0.5 btc | "Lucas"
coin3: 1.0 btc | "David", "Lucas", "Dan"

Which coins are better to select/use?

The current implementation would select all 3 coins, what I am suggesting is we should only select the 3rd coin.

Even if you somehow select only coin1 and coin2 you are leaking some information to David AND to Lucas, while if you use coin3 you are only leaking some information to Dan.

Reply:
> That line still can select pockets that have labels without any relation to the recipient, for example David, Dan.
 
 I am not sure I understand this. If the Recipient is `David, Lucas` and it selects a pocket with `David, Dan` I clearly see there is a relation. Both contain `David`. So this is an acceptable selection.
 
>Let's say the recipient is David, Lucas and you want to send 1 btc and you have the following coins:
> 
> coin1: 0.5 btc | "David"
> coin2: 0.5 btc | "Lucas"
> coin3: 1.0 btc | "David", "Lucas", "Dan"
> 
> Which coins are better to select/use?
>
> The current implementation would select all 3 coins, what I am suggesting is we should only select the 3rd coin.

And yes you are right. In this example David and Lucas is an acceptable selection but `David, Lucas, Dan` is the best selection.

My proposal about the sorting `What do you think about sorting the knownByRecipientPockets by the number of the labels that are found in the recipient?` would exactly do what you want. It would select the pocket that contains the most labels from the recipient labels.

My only concern is this is starting to be more than a UI concept. Initially, we just selected between two pockets. Private / non private, then we distinguished the known / unknown pockets. Which still fits into the UI since the library code has no idea about them. But right now we are trying to select the exact and best coins which normally would be the SmartCoinSelector's task IMO. Maybe I am wrong but it seems we started to mix business logic with UI.

Reply:
Added additional logic.

But in my honest opinion, this is surely mixing business logic with UI and I am sure there are still edge cases when it does not find the best pockets... (or at least I don't know which one should be the best)... I start to feel why Lucas stopped it [here](https://github.com/zkSNACKs/WalletWasabi/pull/2591)

Reply:
I agree this is mixing mixing business logic and UI.

> I start to feel why Lucas stopped it https://github.com/zkSNACKs/WalletWasabi/pull/2591

I get that, but that was not as much important as it is now, because we are doing the selection instead of the user.

My goal with this is to have the minimum obvious use cases, and later on obviously we can improve it more.

Reply:
I improved the logic and I believe it is working correctly now... at least I believe, would be nice to have tests for the most common cases?